include/seqan3/core/math.hpp

```diff
@@ -16,7 +16,80 @@
 #include <seqan3/std/concepts>
 #include <stdexcept>
 
-#include <seqan3/core/platform.hpp>
+#include <seqan3/core/bit_manipulation.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Computes the floor of the logarithm to the base of two for unsigned integers.
+ * \ingroup core
+ * \param[in] n An unsigned integer.
+ * \attention *n = 0* is a special case and is undefined.
+ * \returns \f$ \lfloor log_2(n) \rfloor \f$.
+ *
+ * \details
+ *
+ * The difference to `std::floor(std::log2(n))` is that everything is computed *exactly* (without precision loss due to
+ * promoting to `double`)
+ *
+ * ### Example
+ *
+ * \include test/snippet/core/detail/floor_log2.cpp
+ *
+ * ### Exception
+ *
+ * No-throw guarantee.
+ *
+ * ### Thread-safety
+ *
+ * Thread safe.
+ *
+ * ### Complexity
+ *
+ * Constant.
+ */
+template <std::unsigned_integral unsigned_t>
```

Comment:
so we only allow unsigneds. shouldn't we make a a fallback to `std` for other integer types?

Reply:
`std::log2` return NaN (a special double/float) if the input is smaller than 0 and sets some funny error handling makro or whatever https://en.cppreference.com/w/cpp/numeric/math/log2

So, if we wanted integers we would need error handling since the msb stuff would still work.
And currently, there's no need for it?

Reply:
The problem is that std::log2 is defined on "real" types. So they have to handle negative numbers, the logarithm is "normally" not defined on negative numbers and 0, so limiting this to unsigned types is reasonable.